src/crypto_helper/algorithm.rs

```diff
@@ -361,6 +363,152 @@ pub struct ZlibInput {
     #[serde(serialize_with = "serialize_bytes", deserialize_with = "deserialize_bytes")]
     pub data: Vec<u8>,
 }
+#[derive(Eq, Clone, Copy, PartialEq, Debug, Serialize, Deserialize, Default)]
+pub enum Argon2Variant {
+    Argon2i,
+    Argon2d,
+    #[default]
+    Argon2id,
+}
+#[derive(Eq, Clone, Copy, PartialEq, Debug, Serialize, Deserialize, Default)]
+pub enum Argon2Version {
+    Version10,
+    #[default]
+    Version13,
+}
+#[derive(Eq, Clone, PartialEq, Debug, Serialize, Deserialize)]
+pub struct Argon2HashAction {
+    pub memory: u32,
+    pub iters: u32,
+    pub paralelism: u32,
+    pub output_len: usize,
+    #[serde(serialize_with = "serialize_bytes", deserialize_with = "deserialize_bytes")]
+    pub salt: Vec<u8>,
+    #[serde(serialize_with = "serialize_bytes", deserialize_with = "deserialize_bytes")]
+    pub secret: Vec<u8>,
+    pub variant: Argon2Variant,
+    pub version: Argon2Version,
+}
+
+#[derive(Eq, Clone, PartialEq, Debug, Serialize, Deserialize)]
+pub enum Argon2Action {
+    Hash(Argon2HashAction),
+    Verify(#[serde(serialize_with = "serialize_bytes", deserialize_with = "deserialize_bytes")] Vec<u8>),
+}
+#[derive(Eq, Clone, PartialEq, Debug, Serialize, Deserialize, Default)]
+pub struct Argon2Input {
+    pub action: Argon2Action,
+    #[serde(serialize_with = "serialize_bytes", deserialize_with = "deserialize_bytes")]
+    pub data: Vec<u8>,
+}
+
+impl Argon2Input {
+    pub fn set_variant(&self, variant: Argon2Variant) -> Self {
+        let mut input = self.clone();
+        if let Argon2Action::Hash(ref mut action) = input.action {
+            action.variant = variant;
+        }
+        input
+    }
+    pub fn set_version(&self, version: Argon2Version) -> Self {
+        let mut input = self.clone();
+        if let Argon2Action::Hash(ref mut action) = input.action {
+            action.version = version;
+        }
+        input
+    }
+}
+impl<'a> TryFrom<&'a str> for Argon2Version {
+    type Error = (); // todo: proper error type
+
+    fn try_from(value: &'a str) -> Result<Self, Self::Error> {
+        match value {
+            "Argon13" => Ok(Self::Version13),
+            "Argon10" => Ok(Self::Version10),
+            _ => Err(()),
+        }
+    }
+}
+impl From<Argon2Variant> for argon2::Algorithm {
```

Comment:
add an empty line between impls

```suggestion
}

impl From<Argon2Variant> for argon2::Algorithm {
```